### PR TITLE
Flashcards

### PR DESCRIPTION
**Pull request by omar-mendez1**

### CODE REVIEW

File: client/src/pages/FlashcardsPage.jsx [hunk 1]
```diff
@@ -1,9 +1,310 @@
...
+                </p>
+              </div>
+
+              {/* Answer Choices */}
+              <div className="grid grid-cols-2 gap-4 mt-6">
+                {currentChoices.map((choice, index) => (
+                  <Button
+                    key={index}
+                    className="py-3 px-6 text-lg text-black transition-colors duration-150 bg-[#6AB9F2] rounded-lg focus:shadow-outline hover:bg-[#C6E600] hover:text-[#3A3A3A]"
+                    onClick={() => handleAnswerClick(choice)}
```
Comment (djankies) [inline client/src/pages/FlashcardsPage.jsx:278]:
```suggestion
                    onClick={(e) => handleAnswerClick(choice, e)}
```

File: client/src/pages/FlashcardsPage.jsx [hunk 2]
```diff
@@ -1,9 +1,310 @@
...
+  const shuffle = (array) => {
+    const newArray = [...array];
+    for (let i = newArray.length - 1; i > 0; i--) {
+      const j = Math.floor(Math.random() * (i + 1));
+      [newArray[i], newArray[j]] = [newArray[j], newArray[i]];
+    }
+    return newArray;
+  };
+
+  const handleAnswerClick = (choice) => {
```
Comment (djankies) [inline client/src/pages/FlashcardsPage.jsx:126]:
```suggestion
  const handleAnswerClick = (choice, e) => {
    e.preventDefault();
```

File: client/src/pages/FlashcardsPage.jsx [hunk 3]
```diff
@@ -1,9 +1,310 @@
...
+      setCurrentCardIndex((prev) => prev + 1);
+      setTimeout(() => setCorrectPopUp(false), 1000);
+    } else {
+      setCurrentStreak(0);
+      setIncorrectPopUp(true);
+      setTimeout(() => setIncorrectPopUp(false), 1000);
+    }
+  };
+
+  const handleDeckChange = (deck) => {
```
Comment (djankies) [inline client/src/pages/FlashcardsPage.jsx:143]:
```suggestion
  const handleDeckChange = (deck, e) => {
  e.preventDefault();
```

File: client/src/pages/FlashcardsPage.jsx [hunk 4]
```diff
@@ -1,9 +1,310 @@
...
+    setCurrentDeck(deck);
+    const filteredCards = filterByDiscard(deck.flashcards);
+    setFlashcards(filteredCards);
+    setCurrentCardIndex(0);
+    setNumCorrect(0);
+    setCurrentStreak(0);
+  };
+
+  const handleDiscard = () => {
```
Comment (djankies) [inline client/src/pages/FlashcardsPage.jsx:152]:
```suggestion
  const handleDiscard = (e) => {
    e.preventDefault();
```

File: client/src/pages/FlashcardsPage.jsx [hunk 5]
```diff
@@ -1,9 +1,310 @@
...
+          <div className="skillProgressRing">
+            <p className="text-black">Circle</p>
+          </div>
+          <p className="font-semibold mb-4 text-black">Other Flashcards</p>
+          <div className="flex flex-col gap-2">
+            {themeDecks?.map((deck, index) => (
+              <Button
+                key={index}
+                className="w-full py-2 px-4 text-sm text-[#3A3A3A] transition-colors duration-150 bg-pear rounded-lg focus:shadow-outline hover:bg-[#C6E600]"
+                onClick={() => handleDeckChange(deck)}
```
Comment (djankies) [inline client/src/pages/FlashcardsPage.jsx:300]:
```suggestion
                onClick={(e) => handleDeckChange(deck, e)}
```

File: client/src/pages/FlashcardsPage.jsx [hunk 6]
```diff
@@ -1,9 +1,310 @@
...
+  const [currentCard, setCurrentCard] = useState(null);
+  const [currentChoices, setCurrentChoices] = useState([]);
+
+  // Game state
+  const [numCorrect, setNumCorrect] = useState(0);
+  const [currentStreak, setCurrentStreak] = useState(0);
+  const [correctPopUp, setCorrectPopUp] = useState(false);
+  const [incorrectPopUp, setIncorrectPopUp] = useState(false);
+  const [isCorrect, setIsCorrect] = useState(null); //eslint-disable-line no-unused-vars
+  const [isFinished, setIsFinished] = useState(false);
```
Comment (djankies) [inline client/src/pages/FlashcardsPage.jsx:29]: Better practice to group related variables:
```suggestion
// User-related state
const [userState, setUserState] = useState({
  userData: useUserData()[0], // Assuming useUserData returns a tuple
  language: 'Spanish',
  skillLevel: 'beginner',
  currentTheme: 'family',
});

// Deck and card state
const [deckState, setDeckState] = useState({
  themeDecks: [],
  currentDeck: null,
  flashcards: [],
  currentCardIndex: 0,
  currentCard: null,
  currentChoices: [],
});

// Game state
const [gameState, setGameState] = useState({
  numCorrect: 0,
  currentStreak: 0,
  correctPopUp: false,
  incorrectPopUp: false,
  isCorrect: null, //eslint-disable-line no-unused-vars
  isFinished: false,
});

// Functions to update specific parts of the state
const updateUserState = (updates) => setUserState((prev) => ({ ...prev, ...updates }));
const updateDeckState = (updates) => setDeckState((prev) => ({ ...prev, ...updates }));
const updateGameState = (updates) => setGameState((prev) => ({ ...prev, ...updates }));
```

File: client/src/pages/FlashcardsPage.jsx [hunk 7]
```diff
@@ -1,9 +1,310 @@
...
+  // Update current card when index changes or cards are filtered
+  useEffect(() => {
+    if (flashcards?.length > 0) {
+      if (currentCardIndex < flashcards.length) {
+        setCurrentCard(flashcards[currentCardIndex]);
+        setIsFinished(false);
+      } else {
+        setIsFinished(true);
+      }
+    }
```
Comment (djankies) [inline client/src/pages/FlashcardsPage.jsx:95]:
```suggestion
    if (flashcards?.length > 0 && currentCardIndex < flashcards.length) {
      setCurrentCard(flashcards[currentCardIndex]);
      setIsFinished(false);
    } else {
      setIsFinished(true);
    }
```

File: server/src/index.js
Comment (djankies) [inline server/src/index.js:1]: Language router should be added back in (this file should not have changed)

**Review by djankies — CHANGES_REQUESTED** [review]
It looks good. I added some notes on best practices. Implementing them is unnecessary, but I provided them for your consideration in future updates. One problem, however, needs addressing.

**Issue:** Inadvertently removed a router from index.js that needs to be included in main

File: server/src/index.js
Comment (omar-mendez1) [reply server/src/index.js:1]: added language router back to the file

**Pull request merged**